src/data/vectors/upsert.ts

```diff
@@ -40,16 +41,24 @@ export class UpsertCommand<T extends RecordMetadata = RecordMetadata> {
     });
   };
 
-  async run(records: Array<PineconeRecord<T>>): Promise<void> {
+  async run(
+    records: Array<PineconeRecord<T>>,
+    retryOptions?: RetryOptions
+  ): Promise<void> {
     this.validator(records);
 
     const api = await this.apiProvider.provide();
-    await api.upsertVectors({
+
+    const retryWrapper = new RetryOnServerFailure(
+      api.upsertVectors.bind(api),
```

Comment:
Re: my usage of the `bind` method here: 

From researching a bit, this seemed like the best way to go (as opposed to an arrow function or wrapper) since I made `RetryServerErrors` take a 'function reference' as an argument. With `bind`, I can explicitly bind the `context` (`this`) to the `api` object originally defined as `this` in `index.ts`. 

We need to be explicit about the context of `this` here, since `api` is specific to the context in which it was instantiated (it has a specific user's Pinecone API key and other config options). 

From reading, there's also a non-zero likelihood of the `this` context being lost/detached from its original object in retry functions in particular since they retry the same function over and over again.

Reply:
Thanks for the info, appreciated.